### PR TITLE
luminous: rgw: Multisite sync corruption for large multipart obj

### PR DESCRIPTION
**Pull request by smithfarm**
http://tracker.ceph.com/issues/40146

### CODE REVIEW

**Comment by cbodley** [comment]
jenkins test make check

**Comment by cbodley** [comment]
several conflicts in src/test/rgw/test_http_manager.cc:
- no string_view in c++11
- needs global calls to rgw::curl::setup_curl()/cleanup_curl() and rgw_http_client_curl.h include
- changed interfaces in RGWHTTPClient (constructor arguments, process() call, and expected error code)
```diff
diff --git a/src/test/rgw/test_http_manager.cc b/src/test/rgw/test_http_manager.cc
index 1d292edcc3..60942219ec 100644
--- a/src/test/rgw/test_http_manager.cc
+++ b/src/test/rgw/test_http_manager.cc
@@ -13,6 +13,7 @@
  */
 #include "rgw/rgw_rados.h"
 #include "rgw/rgw_http_client.h"
+#include "rgw/rgw_http_client_curl.h"
 #include "global/global_init.h"
 #include "common/ceph_argparse.h"
 #include <curl/curl.h>
@@ -34,7 +35,7 @@ TEST(HTTPManager, ReadTruncated)
   std::thread server{[&] {
     tcp::socket socket{ioctx};
     acceptor.accept(socket);
-    std::string_view response =
+    std::string response =
         "HTTP/1.1 200 OK\r\n"
         "Content-Length: 1024\r\n"
         "\r\n"
@@ -43,10 +44,14 @@ TEST(HTTPManager, ReadTruncated)
   }};
   const auto url = std::string{"http://127.0.0.1:"} + std::to_string(acceptor.local_endpoint().port());
 
-  RGWHTTPClient client{g_ceph_context, "GET", url};
-  EXPECT_EQ(-EAGAIN, RGWHTTP::process(&client, null_yield));
+  rgw::curl::setup_curl(boost::none);
+
+  RGWHTTPClient client{g_ceph_context};
+  EXPECT_EQ(-EINVAL, client.process("GET", url.c_str()));
 
   server.join();
+
+  rgw::curl::cleanup_curl();
 }
 
 TEST(HTTPManager, SignalThread)
```

**Comment by smithfarm** [comment]
```
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include/boost/asio/impl/write.hpp:25:0,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include/boost/asio/write.hpp:927,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/rgw/test_http_manager.cc:21:
/home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include/boost/asio/detail/consuming_buffers.hpp: In member function ‘boost::asio::detail::consuming_buffers<Buffer, Buffers, Buffer_Iterator>::prepared_buffers_type boost::asio::detail::consuming_buffers<Buffer, Buffers, Buffer_Iterator>::prepare(std::size_t)’:
/home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include/boost/asio/detail/consuming_buffers.hpp:105:50: error: parse error in template argument list
     while (next != end && max_size > 0 && result.count < result.max_buffers)
                                                  ^
src/test/rgw/CMakeFiles/unittest_http_manager.dir/build.make:62: recipe for target 'src/test/rgw/CMakeFiles/unittest_http_manager.dir/test_http_manager.cc.o' failed
make[3]: *** [src/test/rgw/CMakeFiles/unittest_http_manager.dir/test_http_manager.cc.o] Error 1
```

File: src/test/rgw/test_http_manager.cc [hunk 1]
```diff
@@ -13,11 +13,47 @@
  */
 #include "rgw/rgw_rados.h"
 #include "rgw/rgw_http_client.h"
+#include "rgw/rgw_http_client_curl.h"
 #include "global/global_init.h"
 #include "common/ceph_argparse.h"
 #include <curl/curl.h>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/write.hpp>
```
Comment (smithfarm) [inline src/test/rgw/test_http_manager.cc]: @cbodley Could it be that this line is causing the build failure? Odd...

**Comment by smithfarm** [comment]
@cbodley Whelp

**Comment by theanalyst** [comment]
jenkins test this please

**Comment by smithfarm** [comment]
The build error is the same as before:

```
In file included from /home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include/boost/asio/impl/write.hpp:25:0,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include/boost/asio/write.hpp:927,
                 from /home/jenkins-build/build/workspace/ceph-pull-requests/src/test/rgw/test_http_manager.cc:21:
/home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include/boost/asio/detail/consuming_buffers.hpp: In member function 'boost::asio::detail::consuming_buffers<Buffer, Buffers, Buffer_Iterator>::prepared_buffers_type boost::asio::detail::consuming_buffers<Buffer, Buffers, Buffer_Iterator>::prepare(std::size_t)':
/home/jenkins-build/build/workspace/ceph-pull-requests/build/boost/include/boost/asio/detail/consuming_buffers.hpp:105:50: error: parse error in template argument list
     while (next != end && max_size > 0 && result.count < result.max_buffers)
                                                  ^
src/test/rgw/CMakeFiles/unittest_http_manager.dir/build.make:62: recipe for target 'src/test/rgw/CMakeFiles/unittest_http_manager.dir/test_http_manager.cc.o' failed
make[3]: *** [src/test/rgw/CMakeFiles/unittest_http_manager.dir/test_http_manager.cc.o] Error 1
```

**Comment by smithfarm** [comment]
Is this worth trying to salvage for 12.2.13?

**Comment by smithfarm** [comment]
> Is this worth trying to salvage for 12.2.13?

No.

**Pull request closed without merging**